File: javascript/json-transform/src/functions/TransformerFunctionSwitch.ts
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { getAsString, isMap } from "../JsonHelpers";
import { ArgType } from "./common/ArgType";

class TransformerFunctionSwitch extends TransformerFunction {
  constructor() {
    super({
      argsSets: [
        [
          { name: "cases", type: ArgType.Object },
          { name: "default", type: ArgType.Any },
        ],
      ],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const value = await context.getString(null);
    const caseMap = await context.getJsonElement("cases");
    if (!isMap(caseMap)) {
      console.warn("{}.cases was not specified with an object as case map", context.getAlias());
      return null;
    }
    return Object.prototype.hasOwnProperty.call(caseMap, value ?? "null")
      ? caseMap[value ?? "null"]
      : context.getJsonElement("default");
  }
}

export default TransformerFunctionSwitch;

File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionSwitch.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;
import co.nlighten.jsontransform.functions.common.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

public class TransformerFunctionSwitch extends TransformerFunction {
    static final Logger logger = LoggerFactory.getLogger(TransformerFunctionSwitch.class);

    public TransformerFunctionSwitch() {
        super(FunctionDescription.of(
            Map.of(
            "cases", ArgumentType.of(ArgType.Object).position(0),
            "default", ArgumentType.of(ArgType.Any).position(1)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var alias = context.getAlias();
        var value = context.getString(null);
        var adapter = context.getAdapter();
        var caseMap = context.getJsonElement("cases");
        if (!adapter.isJsonObject(caseMap)) {
            logger.warn("{}.cases was not specified with an object as case map", alias);
            return null;
        }
        return adapter.has(caseMap, value)
                ? adapter.get(caseMap, value)
                : context.getJsonElement("default");
    }
}
